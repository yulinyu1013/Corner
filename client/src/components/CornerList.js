import React, { useState, useEffect } from 'react';
import Select from 'react-select';
import { useHistory } from 'react-router-dom';
import '../styles/cornerlist.css';
import { FaSearch, FaSortAmountDown, FaRocketchat } from 'react-icons/fa';
import groupListApis from '../fetchers/groupListApis';
import dummy from '../images/logo.png';

function CornerList({ isExplored, isLoggedIn, currentUser }) {
  const [cornerlist, setCornerlist] = useState([]);
  const [allCorners, setAllCorners] = useState([]);
  const [sortby, setSortBy] = useState('default');
  const [searchTerm, setSearchTerm] = useState('');
  // const [localLoggedIn] = useState(isLoggedIn);

  useEffect(() => {
    if (isLoggedIn && !isExplored) {
      groupListApis.fetchUserGroups(currentUser.id).then((res) => {
        console.log('my corner', isLoggedIn);
        console.log(res);
        setCornerlist(res);
        // console.log(cornerlist);
        setAllCorners(res);
        // setTimeout(() => {}, 1000);
      });
    } else {
      groupListApis.fetchAllPublicGroups().then((res) => {
        console.log('popular corner', isLoggedIn);
        console.log(res);
        setCornerlist(res);
        setAllCorners(res);
        // setTimeout(() => {}, 1000);
      });
    }
  }, [isLoggedIn]);

  useEffect(() => {
    if (searchTerm !== '') {
      const lowSearchTerm = searchTerm.toLowerCase();
      const filteredData = [];
      allCorners.forEach((corner) => {
        corner.tag = corner.tag.map((a) => a.toLowerCase());
        if (corner.tag.includes(lowSearchTerm)) {
          filteredData.push(corner);
          console.log(corner.name);
        }
      });
      setCornerlist(filteredData);
    } else {
      if (isLoggedIn && !isExplored) {
        groupListApis.fetchUserGroups(currentUser.id).then((res) => {
          console.log('my corner2', isLoggedIn);
          console.log(res);
          setCornerlist(res);
          // console.log(cornerlist);
          // setAllCorners(res);
          // setTimeout(() => {}, 1000);
        });
      } else {
        groupListApis.fetchAllPublicGroups().then((res) => {
          console.log('popular corner2', isLoggedIn);
          console.log(res);
          setCornerlist(res);
          // setAllCorners(res);
          // setTimeout(() => {}, 1000);
        });
      }
      setSortBy('default');
    }
  }, [searchTerm, isLoggedIn]);

  const options = [
    // { value: 'message', label: 'message' },
    { value: 'member', label: 'member' },
    { value: 'post', label: 'posts' },
  ];

  const history = useHistory();

  const selectStyles = {
    option: (provided, state) => ({
      ...provided,
      color: state.isSelected ? 'black' : 'grey',
    }),
    control: (base) => ({
      ...base,
      fontFamily: 'Montserrat',
      textAlign: 'start',
      height: 35,
      minHeight: 35,
      width: 170,
      '&:hover': {
        borderColor: 'lightgray',
        boxShadow: '0 0 10px grey',
      }, // border style on hover
      '&:focus': {
        borderColor: 'grey',
      },
      border: '1px solid lightgray', // default border color
    }),
    menu: (base) => ({
      ...base,
      fontFamily: 'Montserrat',
    }),
  };

  /**
   * handle the change on select component
   * @param {} selectedOption
   */
  const handleChange = (selectedOption) => {
    let newCorners = [];
    // if (selectedOption.value === 'message') {
    //   newCorners = [...cornerlist].sort((c1, c2) => {
    //     if (c1.msg === null) {
    //       return 1;
    //     }
    //     if (c2.msg === null) {
    //       return -1;
    //     }
    //     return c2.msg - c1.msg;
    //   });
    //   console.log(newCorners);
    // } else
    if (selectedOption.value === 'member') {
      newCorners = [...cornerlist].sort((c1, c2) => (c2.mems - c1.mems));
    } else {
      newCorners = [...cornerlist].sort((c1, c2) => (c2.posts - c1.posts));
    }
    console.log(selectedOption);
    setCornerlist(newCorners);
    setSortBy(selectedOption.value);
    console.log(newCorners);
    console.log(cornerlist);
  };

  const redirectToCornerPage = (id, name) => {
    history.push(`/corner/${currentUser.name}/${name}/${id}`);
  };

  const redirectToGroupChat = (groupId) => {
    history.push(`/gChat/${groupId}/${currentUser.name}`);
  };

  return (
    <div className="corner-list-wrapper">
      <div className="corner-header">
        <span>{isLoggedIn && !isExplored ? 'My Corner' : 'Popular Corner'}</span>
      </div>
      <div className="sort-search">
        <div className="search-bar">
          <FaSearch className="search-icon" />
          <input
            className="input-field"
            type="text"
            placeholder="Search by Tags..."
            onChange={(e) => {
              setSearchTerm(e.target.value);
            }}
          />
        </div>
        <div className="sort-bar">
          <FaSortAmountDown className="sort-icon" />
          <Select
            value={sortby}
            options={options}
            onChange={handleChange}
            styles={selectStyles}
            placeholder={sortby}
          />
        </div>
      </div>
      <div className="list-container">
        {cornerlist.map(({
          id, name, avatar, posts, mems,
        }) => (
          <div className="singlecorner">
            <div key={id} className="corner" onClick={() => redirectToCornerPage(id, name)} role="presentation">
              <div className="corner-img-container">
                <img className="corner-img" src={avatar || dummy} alt="img" />
              </div>
              <div className="corner-info">
                <div className="corner-name">
                  {name}
                </div>
                <div className="corner-data">
                  {posts}
                  <span>&nbsp;posts|&nbsp;</span>
                  {mems}
                  <span>&nbsp;mems</span>
                </div>
              </div>
            </div>
            <div className="chat-icon" onClick={() => redirectToGroupChat(id, currentUser.name)} role="presentation">
              {(isLoggedIn && !isExplored) && <FaRocketchat className="chat-icon" />}
            </div>
          </div>
        ))}
      </div>
    </div>
  );
}

export default CornerList;
