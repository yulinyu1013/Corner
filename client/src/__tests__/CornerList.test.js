import React from 'react';
import renderer from 'react-test-renderer';
import CornerList from '../components/CornerList';
import user from '../images/user.png';

const currentUser = {
  id: '1',
  name: 'test',
  avatar: {user},
  bio: '',
}

it('renders CornerList correctly when user not loggedin', () => {
  const tree = renderer.create(<CornerList isLoggedIn={false} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders CornerList correctly when user loggedin', () => {
  const tree = renderer.create(<CornerList isLoggedIn currentUser={currentUser} />).toJSON();
  expect(tree).toMatchSnapshot();
});
