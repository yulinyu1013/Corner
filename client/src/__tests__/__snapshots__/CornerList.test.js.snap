// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders CornerList correctly when user loggedin 1`] = `
<div
  className="corner-list-wrapper"
>
  <div
    className="corner-header"
  >
    <span>
      My Corner
    </span>
  </div>
  <div
    className="sort-search"
  >
    <div
      className="search-bar"
    >
      <svg
        className="search-icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
        />
      </svg>
      <input
        className="input-field"
        onChange={[Function]}
        placeholder="Search by Tags..."
        type="text"
      />
    </div>
    <div
      className="sort-bar"
    >
      <svg
        className="sort-icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M304 416h-64a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm-128-64h-48V48a16 16 0 0 0-16-16H80a16 16 0 0 0-16 16v304H16c-14.19 0-21.37 17.24-11.29 27.31l80 96a16 16 0 0 0 22.62 0l80-96C197.35 369.26 190.22 352 176 352zm256-192H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h192a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm-64 128H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h128a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zM496 32H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h256a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
        />
      </svg>
      <div
        className=" css-b62m3t-container"
        onKeyDown={[Function]}
      >
        <span
          className="css-1f43avz-a11yText-A11yText"
          id="react-select-3-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          className="css-1f43avz-a11yText-A11yText"
        />
        <div
          className=" css-1aboltx-control"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className=" css-319lph-ValueContainer"
          >
            <div
              className=" css-14el2xx-placeholder"
              id="react-select-3-placeholder"
            >
              default
            </div>
            <div
              className=" css-14dclt2-Input"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-controls="react-select-3-listbox"
                aria-describedby="react-select-3-placeholder"
                aria-expanded={false}
                aria-haspopup={true}
                aria-owns="react-select-3-listbox"
                autoCapitalize="none"
                autoComplete="off"
                autoCorrect="off"
                className=""
                disabled={false}
                id="react-select-3-input"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                role="combobox"
                spellCheck="false"
                style={
                  Object {
                    "background": 0,
                    "border": 0,
                    "color": "inherit",
                    "font": "inherit",
                    "gridArea": "1 / 2",
                    "label": "input",
                    "margin": 0,
                    "minWidth": "2px",
                    "opacity": 1,
                    "outline": 0,
                    "padding": 0,
                    "width": "100%",
                  }
                }
                tabIndex={0}
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              className=" css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              className=" css-tlfecz-indicatorContainer"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-tj5bde-Svg"
                focusable="false"
                height={20}
                viewBox="0 0 20 20"
                width={20}
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="list-container"
  />
</div>
`;

exports[`renders CornerList correctly when user not loggedin 1`] = `
<div
  className="corner-list-wrapper"
>
  <div
    className="corner-header"
  >
    <span>
      Popular Corner
    </span>
  </div>
  <div
    className="sort-search"
  >
    <div
      className="search-bar"
    >
      <svg
        className="search-icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
        />
      </svg>
      <input
        className="input-field"
        onChange={[Function]}
        placeholder="Search by Tags..."
        type="text"
      />
    </div>
    <div
      className="sort-bar"
    >
      <svg
        className="sort-icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M304 416h-64a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm-128-64h-48V48a16 16 0 0 0-16-16H80a16 16 0 0 0-16 16v304H16c-14.19 0-21.37 17.24-11.29 27.31l80 96a16 16 0 0 0 22.62 0l80-96C197.35 369.26 190.22 352 176 352zm256-192H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h192a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm-64 128H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h128a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zM496 32H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h256a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
        />
      </svg>
      <div
        className=" css-b62m3t-container"
        onKeyDown={[Function]}
      >
        <span
          className="css-1f43avz-a11yText-A11yText"
          id="react-select-2-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          className="css-1f43avz-a11yText-A11yText"
        />
        <div
          className=" css-1aboltx-control"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className=" css-319lph-ValueContainer"
          >
            <div
              className=" css-14el2xx-placeholder"
              id="react-select-2-placeholder"
            >
              default
            </div>
            <div
              className=" css-14dclt2-Input"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-controls="react-select-2-listbox"
                aria-describedby="react-select-2-placeholder"
                aria-expanded={false}
                aria-haspopup={true}
                aria-owns="react-select-2-listbox"
                autoCapitalize="none"
                autoComplete="off"
                autoCorrect="off"
                className=""
                disabled={false}
                id="react-select-2-input"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                role="combobox"
                spellCheck="false"
                style={
                  Object {
                    "background": 0,
                    "border": 0,
                    "color": "inherit",
                    "font": "inherit",
                    "gridArea": "1 / 2",
                    "label": "input",
                    "margin": 0,
                    "minWidth": "2px",
                    "opacity": 1,
                    "outline": 0,
                    "padding": 0,
                    "width": "100%",
                  }
                }
                tabIndex={0}
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              className=" css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              className=" css-tlfecz-indicatorContainer"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-tj5bde-Svg"
                focusable="false"
                height={20}
                viewBox="0 0 20 20"
                width={20}
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="list-container"
  />
</div>
`;
